### DISCUSSION TITLE
Parallel loading from different drives/dirs

### DISCUSSION THREAD

**Discussion by kvaster** [seaweedfs/seaweedfs#3580] [Ideas]
I our setup we have several bare metal servers for seaweedfs volumes. Each server has 10 drives. Each drive provided to seaweedfs as a separate dir: `-dir=/volumes/data01,/volumes/data02,/volumes/data03...`.

I've noticed that loading data files after restart is made in parallell, but seaweedfs processes dir by dir. i.e. seaweedfs loads data from `/volumes/data01` in parallel and after that it goes to next dir.

Is it possible to process dirs in parallel ? This may really speed up starting seaweedfs in such scenario.

**Comment by chrislusf** [comment]
Yes, this could make sense. But it needs some benchmark numbers to prove. The CPU and memory cost may be the bottleneck instead of disk IO since the idx files are not that large.

**Reply by kvaster** [reply]
On my servers cpu load is less then 4% and iowait is 48-50 - cpu and memory will definitely be not an bottleneck. What about an option for parallel dirs processing ?

**Reply by chrislusf** [reply]
Less option is better.

**Reply by kvaster** [reply]
Just took a look on a setup. Each server have 10 drives and each drive has 2TB data files (approx).

Just noticed that we have one option: `-volumeSizeLimitMB=1024`. Could this be a problem ?

**Reply by chrislusf** [reply]
smaller `volumeSizeLimitMB` will create more volumes. The right value depends on each use case.

**Reply by kvaster** [reply]
And could it be a problem why it takes much IO on a start ? Also second question is if it's possible to 'repack' volume data with higher volumeSizeLimitMB without loosing data ....

**Reply by chrislusf** [reply]
I can not tell without statistics.

Merging volumes is not supported.

**Comment by kvaster** [comment]
I've made a small patch for myself:
[concurrency.zip](https://github.com/seaweedfs/seaweedfs/files/9735143/concurrency.zip)

It adds support for changing concurrency for loading volumes and concurrency for loading disks. As I said before my setup is 4x nodes and each node has 10x 18TB hdd drives. Change parameter SEAWEEDFS_VOLUME_CONCURRENCY (defaults to number of cores) make little difference. But changing SEAWEEDFS_DISK_CONCURRENCY to 10 (defaults to 1) decreases startup time by 10x. Also CPU is not more then 5% in both scenarios. So it really makes sense to load indexes on startup concurently from different drives on startup. This is true at least for HDD drives. Behaviour may be (I think it will be) different with NVME drives, but I have no any machine with lot's of nvme drives :)

**Reply by kmlebedev** [reply]
I make PR with patch https://github.com/seaweedfs/seaweedfs/pull/3801

**Reply by kvaster** [reply]
For now it's only discussion. It was easy to test the feature with env variables, but I'm not sure if it's a good way...

**Reply by kmlebedev** [reply]
> For now it's only discussion. It was easy to test the feature with env variables, but I'm not sure if it's a good way...

Yes, it's better to somehow bind this parameter to a disk name or disk type

**Reply by chrislusf** [reply]
We can parallel for all the directories, avoiding this extra parameter?

**Reply by kvaster** [reply]
With HDDs I'm sure it will be OK. I just can't make tests using bunch of nvmes. With nvme's we can be cpu/memory bound. The best implementation probably will be to have some kind of smart queue - which will have ccumulative restriction on in-process indexes which will ensure maximum parallel loading from different disks...

But I've tested parallel loading for all the directories and I see no real impact on cpu or memory. Also I see no real impact even then all indexes are already cached by filesystem. Without being cached it takes 24 seconds to load, when cached it takes 1 second to load. Without parallel loading it takes 240 seconds to load - 24 * number of disks.

My bet is to go with parallel loading without additional parameters.

**Reply by kvaster** [reply]
I will prepare PR :)
